### PR TITLE
Making possible to fill the Google map key

### LINKED ISSUE #27531: Google Maps does not work on the rules page
### Parent Issue

https://github.com/dotCMS/core/issues/14568

### Problem Statement

The Google Maps feature within the visitor location rule on the rules page shows "this page can't load Google Maps correctly"

### Steps to Reproduce

Generate an API key on Google Cloud
Add the API key to the site configuration for dotCMS
Save the configuration
Go to the rules page and add a new rule
Select Visitor's location in the first dropdown
Click +Show Map
The Google Map that is displayed will show the error
Screenshot 2024-02-06 at 1 28 50 PM

### Acceptance Criteria

Allow the rules page to load the Google Maps API with the API key

API Key should be pulled from the Site definition (there is a Google Map Field in the Site Content Type).

### dotCMS Version

5.3.8.14 +
Also happens in most recent qa branch

### Proposed Objective

Technical User Experience

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

https://dotcms.zendesk.com/agent/tickets/110883
https://dotcms.zendesk.com/agent/tickets/114986
https://dotcms.slack.com/archives/CSHTYUR7H/p1681329439715809

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by AndreyDotcms**
- Add a new attribute to the site model
- Based on the current site id, get its google map key.

https://github.com/dotCMS/core/assets/127987858/17b3ebdf-6572-4be8-97fa-51747ad280cc

### CODE REVIEW

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts [hunk 1]
```diff
@@ -26,6 +30,18 @@ export class GoogleMapService {
...
+    //this method gets the Google key from the current sie
+    private getApiKey(): void {
+        const siteId = this.siteService.currentSite.identifier;
+
+        this.siteService
+            .getSiteById(siteId)
+            .pipe(takeUntilDestroyed())
+            .subscribe((site) => {
+                this.apiKey = site.googleMap || null;
+            });
```
[the remark this answers is not in the document]
Comment (oidacra) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:42]: good question. If the site change, Do you need to change the apikey too?

[the remark this answers is not in the document]
Comment (AndreyDotcms) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:42]: Yes, we want to change the apiKey, if the site changes.

[the remark this answers is not in the document]
Comment (hmoreras) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:42]: `this.siteService.currentSite` does not have the `googleMap` prop ?. and as the guys mention we should use `siteService.switchSite$`

[the remark this answers is not in the document]
Comment (AndreyDotcms) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:42]: > `this.siteService.currentSite` does not have the `googleMap` prop ?. and as the guys mention we should use `siteService.switchSite$`

No, `this.siteService.currentSite` does not contain the `googleMap` property. It makes a request to `v1/site/currentSite`, which does not include that attribute. 
I understand the issue and I am working on it.

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts [hunk 2]
```diff
@@ -16,36 +20,59 @@ export class GoogleMapService {
...
-            }
-            this.addScript(url);
+        this.loadingApi = true;
+        let url: string;
+        if (this.apiKey) {
+            url = `https://maps.googleapis.com/maps/api/js?key=${this.apiKey}&callback=mapsApiReady`;
+        } else {
+            url = `https://maps.googleapis.com/maps/api/js?callback=mapsApiReady`;
         }
+        console.log('URL to render', url);
```
Comment (fmontes) [inline core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:57]: remove.

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts [hunk 3]
```diff
@@ -233,4 +233,5 @@ export interface Site {
     type: string;
     identifier: string;
     archived: boolean;
+    googleMap?: string;
```
Comment (fmontes) [inline core-web/libs/dotcms-js/src/lib/core/site.service.ts:236]: Why here? is this prop coming back in the site api?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/dotcms-js/src/lib/core/site.service.ts (hunk 3 above)
Comment (AndreyDotcms) [reply core-web/libs/dotcms-js/src/lib/core/site.service.ts:236]: Yes, it comes from the site api request. When requesting the actual site data, we have to access the `site.googleMap` attribute. If we do not declare the `googleMap` attribute in the interface, we will get the following error.
TS2339: The property `googleMap` does not exist in the `site` type.

Not all site requests include that attribute, that is why I am assigning the `?`.

**Review by fmontes — APPROVED** [review]

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts [hunk 4]
```diff
@@ -11,41 +17,55 @@ window['mapsApiReady'] = () => {
...
-
-    apiReady: Boolean = false;
-    loadingApi: Boolean = false;
+    apiReady = false;
     mapsApi$: BehaviorSubject<{ ready: boolean; error?: any }>;
-
-    constructor() {
+    private destroy$ = new Subject<boolean>();
+    constructor(private siteService: SiteService, private dotSiteService: DotSiteService) {
+        this.loadApi(this.siteService.currentSite.identifier).subscribe(() => {});
```
Comment (hmoreras) [inline core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:24]:
```suggestion
        this.loadApi(this.siteService.currentSite.identifier).subscribe();
```

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts [hunk 5]
```diff
@@ -11,41 +17,55 @@ window['mapsApiReady'] = () => {
...
                 this.apiReady = true;
             }
         });
+
+        this.siteService.switchSite$
+            .pipe(
+                takeUntil(this.destroy$),
+                switchMap(({ identifier }) => this.loadApi(identifier))
+            )
+            .subscribe(() => {});
```
Comment (hmoreras) [inline core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:37]:
```suggestion
            .subscribe();
```

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts [hunk 6]
```diff
@@ -11,41 +17,55 @@ window['mapsApiReady'] = () => {
 
 @Injectable()
 export class GoogleMapService {
-    apiKey: string = null;
-
-    apiReady: Boolean = false;
-    loadingApi: Boolean = false;
+    apiReady = false;
```
Comment (hmoreras) [inline core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts:20]: check if we can remove this variable.

**Review by oidacra — APPROVED** [review]

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts (hunk 5 above)
Comment (AndreyDotcms) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts]: Done

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts (hunk 4 above)
Comment (AndreyDotcms) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts]: Done

File: core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts (hunk 6 above)
Comment (AndreyDotcms) [reply core-web/libs/dot-rules/src/lib/services/GoogleMapService.ts]: Done

**Pull request merged**